src/integer.h

```diff
@@ -55,16 +55,26 @@ GIT_INLINE(bool) git__add_uint64_overflow(uint64_t *out, uint64_t one, uint64_t
 }
 
 /* Use clang/gcc compiler intrinsics whenever possible */
-#if (SIZE_MAX == ULONG_MAX) && __has_builtin(__builtin_uaddl_overflow)
-# define git__add_sizet_overflow(out, one, two) \
-	__builtin_uaddl_overflow(one, two, out)
-# define git__multiply_sizet_overflow(out, one, two) \
-	__builtin_umull_overflow(one, two, out)
-#elif (SIZE_MAX == UINT_MAX) && __has_builtin(__builtin_uadd_overflow)
-# define git__add_sizet_overflow(out, one, two) \
-	__builtin_uadd_overflow(one, two, out)
-# define git__multiply_sizet_overflow(out, one, two) \
-	__builtin_umul_overflow(one, two, out)
+#if (__has_builtin(__builtin_add_overflow) || \
+     (defined(__GNUC__) && (__GNUC__ >= 5)))
+
+#	if (ULONG_MAX == ULLONG_MAX) && defined(_WIN64)
+#		define git__add_sizet_overflow(out, one, two) \
+			__builtin_uaddll_overflow(one, two, out)
+#		define git__multiply_sizet_overflow(out, one, two) \
+			__builtin_umulll_overflow(one, two, out)
+#	elif (ULONG_MAX == ULLONG_MAX)
+#		define git__add_sizet_overflow(out, one, two) \
+			__builtin_uaddl_overflow(one, two, out)
+#		define git__multiply_sizet_overflow(out, one, two) \
+			__builtin_umull_overflow(one, two, out)
+#	else
```

Comment:
We're dropping the `SIZE_MAX == UINT_MAX` test.  I'm not sure that there's a guarantee that they're identical in the C standard.  [`size_t` is](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf):

> the unsigned integer type of the result of the sizeof operator

and is a minimum of a 16-bits.  But it doesn't necessarily hold that `size_t` is equivalent to `unsigned int` that I can see in the spec.  I realize that we're unlikely to encounter a system that has 16 bit `size_t`s and 32 bit `unsigned int`s, but without a strict assurance of that from the spec, since this macro is explicitly for security sensitive code, I would like to keep the explicit test.

Reply:
what happens in the final case is that we call the `__builtin_add_overflow` intrinsic, which takes _any_ type you give it. i just added the two other cases explicitly to make clang prefer the other two and avoid a dependency to compiler-rt.

Reply:
Ah, thanks - I had indeed read this as `__builtin_uadd_overflow`.